### PR TITLE
[Swift] Add Strings and Characters concept exercise

### PR DESCRIPTION
**Pull request by wneumann**
Adding basic Strings and Characters exercise to break cycles in the progression graph.

### CODE REVIEW

File: languages/swift/exercises/concept/strings-and-characters/.docs/hints.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+## 1. Create a set of useful strings
+
+- [Strings][strings-and-characters] are defined as literals by enclosing zero or more of characters between double quotation marks (`"`).
+
+## 2. Create a set of useful characters
+
+- [Characters][strings-and-characters] are defined as literals by enclosing exactly one character between double quotation marks (`"`).
+- Characters defined as literals need to be explicitly given a `Character` type annotation to prevent them from being interpreted as Strings by the compiler.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/hints.md:8]: Off-topic: what an interesting approach to characters!

File: languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+// => true
+```
+
+### Iterating over the characters of a string
+
+The individual characters in a string can be stepped through one at a time using a for-in loop. This type of loop takes each character in the string, in order, and binds the character to a specified name for further processing inside the loop body. For example, to print out all of the capital letters in a string one can write:
+
+```swift
+for char in "Hello Борис. How is Наташа?" {
+  if char.isUppercase {
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md:101]: This exercise does not have a prerequisite on `if` statements, does it? As such, you unfortunately can't use it in this exercise unless you want to add it as a dependency.

File: languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+// Swift infers this to be a String
+
+var aChar: Character = "A"
+// Swift now knows to make this a Character
+
+let badChar: Character = "Too many characters"
+// Error: Cannot convert value of type 'String' to specified type 'Character'
+```
+
+Like other constructs in Swift, characters and strings are immutable if defined using `let` and mutable if defined using `var`. So in the above examples, `hello` and `sausage` are immutable strings while `fam` and `empty`, AND `aChar` are mutable.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md:28]: Is this line needed? It sounds like everything works just like the students would expect after having done the `basics` exercise, so I feel that this probably does not add any value here (you could include it in the `after.md` document if you want to emphasize this fact).

File: languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+empty.isEmpty
+// => true
+fam.count
+// => 60
+aChar.isLowercase
+// => false
+let half: Character = "½"
+half.isNumber
+// => true
+```
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md:93]: I like the beginning of this explanation. I would suggest to make make the text be a bit shorter, e.g. by using less examples. The general point that you want to get across to students is that there are such things as properties, that you can access those using dot-notation and that strings and characters have properties. 

The after.md document can be a bit more expansive though.

File: languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+### Iterating over the characters of a string
+
+The individual characters in a string can be stepped through one at a time using a for-in loop. This type of loop takes each character in the string, in order, and binds the character to a specified name for further processing inside the loop body. For example, to print out all of the capital letters in a string one can write:
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md:97]: This exercise introduces a `for-in` loop. I would suggest adding this to the `concepts` unlocked by the exercise, as I assume this is something you want to depend on later?

File: languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+Strings and Characters can be compared for equality (or lack thereof) using the `==` operator (or `!=` for not-equal).
+
+```swift
+let hi = "Hello"
+let period: Character = "."
+"\(hi)\(period)" == "Hello."
+// => true
+
+period != ";"
+// => true
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md:74]: The string interpolation made this example somewhat harder to read to me, as a non-Swift programmer. Maybe also shorten the example a bit:

```suggestion
let hi = "Hello"
hi == "Hello"
// => true

let period: Character = "."
period != ";"
// => true
```

File: languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
+Strings in Swift are a collection of characters, where characters are, more or less, a single printable element. So strings in Swift are able to contain Unicode characters and emoji.
+
+Strings are easily created in Swift through string literals. String literals are similar to those seen in many other programming languages, characters enclosed between a pair of double quotation marks (`"`). String literals can include the following special characters:
+\\0 (null character), \\\\ (backslash), \\t (horizontal tab), \\n (newline), \\r (carriage return), \\" (double quotation mark) and \\' (single quotation mark).
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md:4]:
```suggestion
`\0` (null character), `\\` (backslash), `\t` (horizontal tab), `\n` (newline), `\r` (carriage return), `\"` (double quotation mark) and `\'` (single quotation mark).
```

Comment (ErikSchierboom) [reply languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md:4]: The introduction.md document doesn't have to be exhaustive, it should give the student just enough context to figure out the solution themselves. As such, you could consider shortening this a bit by mentioning that certain characters require escaping by prefixing them with the `\` character, and then naming one example. The after.md can then be exhaustive.

File: languages/swift/exercises/concept/strings-and-characters/.docs/instructions.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+Implement the function `graduationFor(name: String, year: Int) -> String` which takes a name as a string parameter and a year as a string parameter and uses string interpolation to create a phrase for a sign that uses a newline to separate the two lines of the message.
+
+```swift
+graduationFor(name: "Padma", year: 2020)
+// => "Congratulations Padma!\nClass of 2020"
+```
+
+## 5. Count the number of lines in a sign's message
+
+The signs that the messages go on can only hold up to four lines. Implement the function `numberOfLines(in: String) -> Int` that takes a string as input, iterates over the characters in the string to count the number of newline characters. Return the number of newlines plus 1 as the number of lines in the message.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/instructions.md:41]: Could you add an example?

File: languages/swift/exercises/concept/strings-and-characters/.docs/instructions.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+## 6. Count the number of numbers in a sign's message
+
+Because they are lower in stock, numbers cost more to put in a sign's message than letters. Implement a function `numberOfNumbers(in: String) -> Int` that takes a string as input, iterates over the characters in the string to count the number of characters which are numbers and return this count.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/instructions.md:45]: Could you add an example?

File: languages/swift/exercises/concept/strings-and-characters/.docs/hints.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+- Characters cannot be concatenated with Strings. They must first be converted into Strings.
+
+## 4. Build a graduation sign
+
+- [String interpolation][string-interpolation] is performed by placing the expression whose value you want to place in the string inside a set of parentheses which are preceded by a backslash character.
+
+## 5. Count the number of lines in a sign's message
+
+- One can iterate over the characters of a string using a for-in loop.
+- Characters and strings can be [compared for equality][string-equality] with the `==` operator.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/hints.md:22]: This line gives away the solution (using the `==` operator). Maybe you could rephrase to not give away the solution directly?

File: languages/swift/exercises/concept/strings-and-characters/.docs/hints.md [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+- [Strings][strings-and-characters] are defined as literals by enclosing zero or more of characters between double quotation marks (`"`).
+
+## 2. Create a set of useful characters
+
+- [Characters][strings-and-characters] are defined as literals by enclosing exactly one character between double quotation marks (`"`).
+- Characters defined as literals need to be explicitly given a `Character` type annotation to prevent them from being interpreted as Strings by the compiler.
+
+## 3. Combine phrases to build up messages
+
+- Strings can be concatenated using the `+` operator.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/hints.md:12]: This line gives away the solution (using the `+` operator). Maybe you could rephrase and link to an external resource to not give away the solution directly?

File: languages/swift/exercises/concept/strings-and-characters/.docs/after.md [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+// prints out:
+// H
+// Б
+// H
+// Н
+```
+
+## NSString
+
+Strings in Swift also obtain all of the functionality of the [NSString][nsstring-docs] class which was originally used in Apple's Objective-C libraries.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/after.md:180]: Good addition.

File: languages/swift/exercises/concept/strings-and-characters/.docs/after.md [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+### Type conversion
+
+Type conversion allows us to get the string representation of many different types in Swift, including `Int`s, `Double`s and `Character`s.
+
+```swift
+String(110)
+// => "110"
+String(Double.pi)
+// => "3.141592653589793"
+String("x" as Character)
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/after.md:39]: The `as` operator is not something the student will be familiar with if I checked the prerequisites. Maybe rewrite this to use a variable instead?

File: languages/swift/exercises/concept/strings-and-characters/.docs/after.md [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+
+let areYouHappy = "Happy" + String(question)
+// => "Happy?"
+```
+
+Strings may also be built up out of values of other types in two other ways: format strings, or string interpolation.
+
+### Format strings
+
+Format strings are similar to those seen in other languages, where a [string with special tokens inside][string-format-specifiers] is followed by a list of components that are inserted into the string at the location of the tokens, formatted as directed by the tokens:
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/after.md:99]: Just a FYI: some tracks have a separate exercise for working with string using more complex formatting stuff.

File: languages/swift/exercises/concept/strings-and-characters/.meta/Example.swift [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+func graduationFor(name: String, year: Int) -> String {
+  "Congratulations \(name)!\nClass of \(year)"
+}
+
+func numberOfLines(in message: String) -> Int {
+  var newlineCount = 1
+  for char in message {
+    if char == "\n" { newlineCount += 1 }
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.meta/Example.swift:19]: See earlier comment on using an `if` without it being in the exercise's prerequisites.

File: languages/swift/exercises/concept/strings-and-characters/.meta/Example.swift [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+  }
+  return newlineCount
+}
+
+func numberOfNumbers(in message: String) -> Int {
+  var numberCount = 0
+  for char in message {
+    if char.isNumber { numberCount += 1 }
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.meta/Example.swift:27]: See earlier comment on using an `if` without it being in the exercise's prerequisites.

File: languages/swift/exercises/concept/strings-and-characters/.meta/design.md [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+- String indexing
+- String decomposition
+- String conformance to `Collection`.
+- String slicing/Substrings.
+- Multi-line strings.
+- Raw strings.
+- Memory and performance characteristics.
+- Character views, utf8 views, utf16 views, etc.
+
+## Concepts
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.meta/design.md:29]: It looks like this exercise also introduces the concept of a `for-in` loop.

**Review by ErikSchierboom — APPROVED** [review]

File: languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md (hunk 2 above)
Comment (wneumann) [reply languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md:101]: Doh! Yeah, that was the reason I broke out this exercise - to break the prerequisite loops. I am removing both the for-in and the conditionals portion of this exercise as the aren't needed here and are covered in more depth in following exercises.

File: languages/swift/exercises/concept/strings-and-characters/.docs/after.md (hunk 13 above)
Comment (wneumann) [reply languages/swift/exercises/concept/strings-and-characters/.docs/after.md:39]: Yeah, I'll remove this and cover it in a type conversion exercise down the road after classes are introduced.

File: languages/swift/exercises/concept/strings-and-characters/.docs/after.md (hunk 14 above)
Comment (wneumann) [reply languages/swift/exercises/concept/strings-and-characters/.docs/after.md:99]: I may add an exercise like that, though in Swift they aren't used much. You're more likely to see some combo of string interpolation, number formatters, date formatters, etc. Though each of those can use an exercise down the road.

File: languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md (hunk 5 above)
Comment (wneumann) [reply languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md:97]: I will be removing this part as there are no good motivating examples without the conditionals, plus it is covered in more depth in the `loops` exercise a couple of steps up the road.

**Comment by wneumann** [comment]
Pushed a new commit to address these comments. The biggest change was removing the iterating over a string section which introduced for-in loops and if statements, which are covered in later exercises. They were replaced with a task involving string properties.

File: languages/swift/exercises/concept/strings-and-characters/.docs/after.md (hunk 14 above)
Comment (ErikSchierboom) [reply languages/swift/exercises/concept/strings-and-characters/.docs/after.md:100]: Okay great

File: languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md (hunk 5 above)
Comment (ErikSchierboom) [reply languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md]: Excellent

File: languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md [hunk 18]
```diff
@@ -0,0 +1,85 @@
+Strings in Swift are a collection of characters, where characters are, more or less, a single printable element. So strings in Swift are able to contain Unicode characters and emoji.
+
+Strings are easily created in Swift through string literals. String literals are similar to those seen in many other programming languages, characters enclosed between a pair of double quotation marks (`"`). Certain characters use a special notation inside of strings so they can be properly represented, such as the newline character which is written as `\n` inside strings or the double quote character which is written as `\"` in of strings so the compiler knows it doesn't end the string.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md:3]: "in of strings" is this missing a word?

File: languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+// => true
+
+let period: Character = "."
+period != ";"
+// => true
+```
+
+### String and Character properties
+
+Strings and characters have a few different _properties_ which can be queried to get information about the string or character. These are queried by placing a ._propertyName_ after the string or character in question For example, one can check to see if a string is empty by querying it's `isEmpty` property, and the count of characters in a string can be retrieved using its `count` property.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/strings-and-characters/.docs/introduction.md:76]:
```suggestion
Strings and characters have a few different _properties_ which can be queried to get information about the string or character. These are queried by placing a ._propertyName_ after the string or character in question. For example, one can check to see if a string is empty by querying its `isEmpty` property, and the count of characters in a string can be retrieved using its `count` property.
```

**Review by ErikSchierboom — APPROVED** [review]
Great!

**Review by ErikSchierboom — APPROVED** [review]

**Pull request merged**